### PR TITLE
F 491 implement project verification epic

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
related to Giveth/giveth-dapps-v2#712, Giveth/giveth-dapps-v2#491

### CODE REVIEW

File: src/resolvers/projectVerificationFormResolver.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      return createProjectVerificationRequest({
+        projectId,
+        userId,
+      });
+    } catch (e) {
+      logger.error('createProjectVerificationRequest() error', e);
+      throw e;
+    }
+  }
+}
```
Comment (mohammadranjbarz) [inline src/resolvers/projectVerificationFormResolver.ts:61]: @FatemeGhasemi 
Please write some test cases here for these failed scenarios:
- [ ] When project is already verified and we throw error
- [ ] When projectId is not valid and can't find the project
- [ ] When user is not authenticated
- [ ] When user is not authorized (is not project's owner)
- [ ] When user already have an inprogress project verification form ( you should write two test cases for both `draft` and `submitted` statuses
And anything that can write test for that ( please see the codes and find possible errors and write test for them

**Comment by mohammadranjbarz** [comment]
@FatemeGhasemi 
Please write some test cases here for these failed scenarios for :
* createProjectVerificationForm mutation
- [x] When project is already verified and we throw error
- [x] When projectId is not valid and can't find the project
- [x] When user is not authenticated
- [x] When user is not authorized (is not project's owner)
- [x] When user already have an inprogress project verification form ( you should write two test cases for both `draft` and `submitted` statuses
* getCurrentProjectVerificationForm query:
- [x] When project is already verified and we throw error
- [x] When projectId is not valid and can't find the project
- [x] When user is not authenticated
- [x] When user is not authorized (is not project's owner)
- [x] There is project verification form with status different that `draft` and `submitted`
- [x] When there is a project but it doesn't have any project verification form

And you can write more test cases in projectVerificationRepository.test.ts as well

And anything that can write test for that ( please see the codes and find possible errors and write test for them

**Comment by mohammadranjbarz** [comment]
@CarlosQ96 
Sorry man I couldn't finish this PR,
The main problem that I couldn't fix on this PR is that when I want to get `jsonb` fields in output, I get error from graphql, I didn't know how to fix that, other things seems to be fine ( except updating **managingFunds** step) and the test cases are written but failed scenario for `updateProjectVerificationForm` is not written

After finishing this PR we should tell the frontend guys to how use below queries/mutations
* **getCurrentProjectVerificationForm**
* **createProjectVerificationForm**
* **updateProjectVerificationForm**

File: test/graphqlQueries.ts [hunk 2]
```diff
@@ -781,3 +781,92 @@ export const editProjectUpdateQuery = `
...
+//                       description
+//                       relatedAddress {
+//                         address
+//                         networkId
+//                         title
+//                       }
+//                     }
+//
+//             }
+//         `;
```
Comment (mohammadranjbarz) [inline test/graphqlQueries.ts]: @CarlosQ96 
I mean this, when you want to get `managingFunds` or `milestones` or `projectRegistry` or `projectContacts` you get `400` error

**Comment by mohammadranjbarz** [comment]
> @CarlosQ96 Sorry man I couldn't finish this PR, The main problem that I couldn't fix on this PR is that when I want to get `jsonb` fields in output, I get error from graphql, I didn't know how to fix that, other things seems to be fine ( except updating **managingFunds** step) and the test cases are written but failed scenario for `updateProjectVerificationForm` is not written
> 
> After finishing this PR we should tell the frontend guys to how use below queries/mutations
> 
> * **getCurrentProjectVerificationForm**
> * **createProjectVerificationForm**
> * **updateProjectVerificationForm**

@CarlosQ96  I fixed the updating **managingFunds** step

**Comment by CarlosQ96** [comment]
> > @CarlosQ96 Sorry man I couldn't finish this PR, The main problem that I couldn't fix on this PR is that when I want to get `jsonb` fields in output, I get error from graphql, I didn't know how to fix that, other things seems to be fine ( except updating **managingFunds** step) and the test cases are written but failed scenario for `updateProjectVerificationForm` is not written
> > After finishing this PR we should tell the frontend guys to how use below queries/mutations
> > 
> > * **getCurrentProjectVerificationForm**
> > * **createProjectVerificationForm**
> > * **updateProjectVerificationForm**
> 
> @CarlosQ96 I fixed the updating **managingFunds** step

Thanks Mohammad Ill finish the missing tests and review.

**Comment by mohammadranjbarz** [comment]
@CarlosQ96  I just fixed the problems that I said above  so it's ready for review ( just failed scenario is missing but I think we can write those tests later

**Comment by mohammadranjbarz** [comment]
@CarlosQ96  
@RamRamez 
@MohammadPCh 
@mateodaza 
@alireza-sharifpour 

I dont know who wants to implement project verification pages so I mentioned all the frontend guys 
This is the steps you should do

### Create a project verification form  
```
  mutation createProjectVerificationForm($slug: String!){
           createProjectVerificationForm(slug: $slug) {
                    id
                    isTermAndConditionsAccepted
                    emailConfirmationToken
                    emailConfirmationSent
                    emailConfirmationSentAt
                    emailConfirmedAt
                    emailConfirmed
                    projectRegistry {
                      organizationDescription
                      isNonProfitOrganization
                      organizationCountry
                      organizationWebsite
                    }
                    projectContacts {
                      youtube
                      instagram
                      linkedin
                      facebook
                      instagram
                      twitter
                    }
                    milestones {
                      mission
                      foundationDate
                      achievedMilestones
                      achievedMilestonesProof
                    }
                    managingFunds {
                      description
                      relatedAddresses {
                        address
                        networkId
                        title
                      }
                    }
                    user {
                      id
                      walletAddress
                    }
                    project {
                      id
                      slug
                    }
                    status
                    }
                    
            }
```
* The project should not be `verified` to can create new verification form 
* If there is a verification form with status `submitted` or `draft` user is not allowed to create new one
* User should be the project owner to be able to create verification form

### Get project verification form
```
 query getCurrentProjectVerificationForm($slug: String!){
           getCurrentProjectVerificationForm(slug: $slug) {
                    id
                    isTermAndConditionsAccepted
                    emailConfirmationToken
                    emailConfirmationSent
                    emailConfirmationSentAt
                    emailConfirmedAt
                    emailConfirmed
                    projectRegistry {
                      organizationDescription
                      isNonProfitOrganization
                      organizationCountry
                      organizationWebsite
                    }
                    projectContacts {
                      youtube
                      instagram
                      linkedin
                      facebook
                      instagram
                      twitter
                    }
                    milestones {
                      mission
                      foundationDate
                      achievedMilestones
                      achievedMilestonesProof
                    }
                    managingFunds {
                      description
                      relatedAddresses {
                        address
                        networkId
                        title
                      }
                    }
                    user {
                      id
                      walletAddress
                    }
                    project {
                      id
                      slug
                    }
                    status
                    }
                    
            }
```
You can use this to know what is the state of verification form , ....
* User should be logged-in
* User should be the owner of project

### Update project verification form 
```
        mutation updateProjectVerificationForm($projectVerificationUpdateInput: ProjectVerificationUpdateInput!){
           updateProjectVerificationForm(projectVerificationUpdateInput: $projectVerificationUpdateInput) {
                    id
                    isTermAndConditionsAccepted
                    emailConfirmationToken
                    emailConfirmationSent
                    emailConfirmationSentAt
                    emailConfirmedAt
                    emailConfirmed
                    projectRegistry {
                      organizationDescription
                      isNonProfitOrganization
                      organizationCountry
                      organizationWebsite
                    }
                    projectContacts {
                      youtube
                      instagram
                      linkedin
                      facebook
                      instagram
                      twitter
                    }
                    milestones {
                      mission
                      foundationDate
                      achievedMilestones
                      achievedMilestonesProof
                    }
                    managingFunds {
                      description
                      relatedAddresses {
                        address
                        networkId
                        title
                      }
                    }
                    user {
                      id
                      walletAddress
                    }
                    project {
                      id
                      slug
                    }
                    status
                    }
                    
            }
```
* Just project owner can update verification form
* In every update user should send that step payload
    * `personalInfo`: (email and wallet address are auto filed from logged in user and uneditable)
    ```
    step: 'personalInfo',
    personalInfo : {
            email: 'test@example.com',
            walletAddress: 'xxxxxxxx',
            fullName: 'Carlos Q',
  }
    ``` 
    * `projectContacts`:
    ```
   {
    step: 'projectContacts',
     projectContacts : {
            facebook: 'facebookAddress',
            instagram: 'instagramAddress',
            linkedin: 'linkedinAddress',
            twitter: '',
            youtube: 'youtubeAddress',
  }
    ```
    * `projectRegistry`
    ```
    {
       step: 'projectRegistry',
         projectRegistry :  {
            organizationWebsite: 'org website',
            organizationCountry: 'France',
            isNonProfitOrganization: true,
            organizationDescription: '',
          }
    } 
    ```
    * `managingFunds`
    ```
    {
      step:'managingFunds',
     managingFunds:  {
        description: 'description!!!',
        relatedAddresses: [
          {
            address: generateRandomEtheriumAddress(),
            networkId: NETWORK_IDS.MAIN_NET,
            title: 'test title',
          },
        ]
    }
    ```
    * `milestones`
    ```
     {
        step: 'milestones',
       milestones:  {
          mission: 'mission',
          achievedMilestonesProof: 'an ipfs hash',
          achievedMilestones: 'lots of work',
          foundationDate: new Date(),
        }
     }
    ```
    * `termAndCondition`
    ```
    {
      step:'termAndCondition',
      isTermAndConditionsAccepted: true,
    }
    ```
    * `submit`:
 
 ``` 
   {
      step: 'submit',
    }  
 ```
    
After this step frontend should not do anything else, and all the things should be handled in admin bro

**PS** You can check https://github.com/Giveth/impact-graph/blob/4cb85b2efb9cde7d00bf8a83286a54ac1866f84e/src/resolvers/projectVerificationFormResolver.test.ts to see test cases and know better how you should call the queries/mutations

**Comment by CarlosQ96** [comment]
Added project verification error when fetching it, when it's already verified.

**Review by CarlosQ96 — APPROVED** [review]
Great job Mohammad I think its fine. However Ill be adding the email confirmation part that we talked today. 

The adminbro will be done in another PR, but will merge this one to unblock frontend.

**Comment by CarlosQ96** [comment]
Added two mutations:

Token of the email confirmation have 1 hour expiration. 

- Send Confirmation Email Mutation (response should return the verification form object with emailConfirmedSent and emailConfirmationSentAt, emailConfirmationToken Filled with data).
```
mutation projectVerificationSendEmailConfirmation($projectVerificationFormId: Float!){
          projectVerificationSendEmailConfirmation(projectVerificationFormId: $projectVerificationFormId) {
            id
            isTermAndConditionsAccepted
            emailConfirmationToken
            emailConfirmationSent
            emailConfirmationSentAt
            emailConfirmedAt
            emailConfirmed
            projectRegistry {
              organizationDescription
              isNonProfitOrganization
              organizationCountry
              organizationWebsite
            }
            personalInfo {
              email
              walletAddress
              fullName
            }
            projectContacts {
              youtube
              instagram
              linkedin
              facebook
              instagram
              twitter
            }
            milestones {
              mission
              foundationDate
              achievedMilestones
              achievedMilestonesProof
            }
            managingFunds {
              description
              relatedAddresses {
                address
                networkId
                title
              }
            }
            user {
              id
              walletAddress
            }
            project {
              id
              slug
            }
            status
          }
        }

```
- Confirm Email Mutation (will return the verification form object or an error, the emailConfirmed will change to true)
```
mutation projectVerificationConfirmEmail($emailConfirmationToken: String!){
          projectVerificationConfirmEmail(emailConfirmationToken: $emailConfirmationToken) {
            id
            isTermAndConditionsAccepted
            emailConfirmationToken
            emailConfirmationSent
            emailConfirmationSentAt
            emailConfirmedAt
            emailConfirmed
            projectRegistry {
              organizationDescription
              isNonProfitOrganization
              organizationCountry
              organizationWebsite
            }
            personalInfo {
              email
              walletAddress
              fullName
            }
            projectContacts {
              youtube
              instagram
              linkedin
              facebook
              instagram
              twitter
            }
            milestones {
              mission
              foundationDate
              achievedMilestones
              achievedMilestonesProof
            }
            managingFunds {
              description
              relatedAddresses {
                address
                networkId
                title
              }
            }
            user {
              id
              walletAddress
            }
            project {
              id
              slug
            }
            status
          }
        }
```

**Comment by CarlosQ96** [comment]
@CarlosQ96
@RamRamez
@MohammadPCh
@mateodaza
@alireza-sharifpour
@MoeNick 
Updated mohammad comments and added the email confirmation mutation comments. Will merge

**Comment by CarlosQ96** [comment]
Related to https://github.com/Giveth/giveth-dapps-v2/issues/804

**Comment by MohammadPCh** [comment]
@mohammadranjbarz @CarlosQ96 
Thank you guys :)

**Comment by alireza-sharifpour** [comment]
Thanks @mohammadranjbarz and @CarlosQ96 for the implementation.
For the organizationCountry List, do you provide it or we should hard code countries names?

**Comment by MoeNick** [comment]
> Thanks @mohammadranjbarz and @CarlosQ96 for the implementation. For the organizationCountry List, do you provide it or we should hard code countries names?

Let's make it on the backend to make it standard for scaling: I created the issue here: 
https://github.com/Giveth/impact-graph/issues/522
@mohammadranjbarz @CarlosQ96

**Comment by alireza-sharifpour** [comment]
I read the comments but couldn’t understand two things:
1- The autosave feature should be on the frontend side or the backend?
2- Should we send the filled data step by step or at the end?
@mohammadranjbarz @CarlosQ96 @MoeNick

**Comment by MoeNick** [comment]
> I read the comments but couldn’t understand two things: 1- The autosave feature should be on the frontend side or the backend? 2- Should we send the filled data step by step or at the end? @mohammadranjbarz @CarlosQ96 @MoeNick

In each step, you handle it on the frontend, but at the end of each step you update it in the backend and save it there, does it make sense?

**Comment by alireza-sharifpour** [comment]
> > I read the comments but couldn’t understand two things: 1- The autosave feature should be on the frontend side or the backend? 2- Should we send the filled data step by step or at the end? @mohammadranjbarz @CarlosQ96 @MoeNick
> 
> In each step, you handle it on the frontend, but at the end of each step you update it in the backend and save it there, does it make sense?

It may sometimes conflicts, for example in some scenarios maybe the backend data be different from the data we have in local storage, what are we going to do in these cases? we should show the user the backend data or the localStorage data?
@MoeNick

**Comment by MoeNick** [comment]
> > > I read the comments but couldn’t understand two things: 1- The autosave feature should be on the frontend side or the backend? 2- Should we send the filled data step by step or at the end? @mohammadranjbarz @CarlosQ96 @MoeNick
> > 
> > 
> > In each step, you handle it on the frontend, but at the end of each step you update it in the backend and save it there, does it make sense?
> 
> It may sometimes conflicts, for example in some scenarios maybe the backend data be different from the data we have in local storage, what are we going to do in these cases? we should show the user the backend data or the localStorage data? @MoeNick

you load based on backed data, when user start providing data in each step (only the currant step he's trying to put data) we save it on local, we submit the form fro each step based on storage data (we only update the specific step data to backend services. )

**Pull request merged**